File: workflow/scripts/R/qc_ambient.R
# Stage 02 — ambient RNA removal in R (decontX; SoupX falls back to decontX since the
# GEO matrices are already filtered and SoupX needs raw droplets). Reads/writes .h5ad
# via zellkonverter. Robust to degenerate inputs (passthrough on error).
suppressMessages({
  library(zellkonverter)
  library(SingleCellExperiment)
  library(decontX)
})

log_con <- file(snakemake@log[[1]], open = "wt")
sink(log_con, type = "message")

infile <- snakemake@input[["h5ad"]]
outfile <- snakemake@output[["h5ad"]]
method <- snakemake@params[["method"]]

if (identical(method, "soupx")) {
  message("SoupX needs raw droplets; GEO ships filtered matrices -> using decontX instead.")
}

sce <- readH5AD(infile)
counts_mat <- assay(sce, 1)

ok <- tryCatch({
  res <- decontX(as(counts_mat, "CsparseMatrix"))
  dec <- round(res$decontXcounts)
  assay(sce, 1) <- dec
  sce$ambient_fraction <- res$contamination
  TRUE
}, error = function(e) {
  message("decontX failed, passing counts through: ", conditionMessage(e))
  sce$ambient_fraction <- 0
  FALSE
})

assayNames(sce)[1] <- "X"
writeH5AD(sce, outfile, X_name = "X")
message(sprintf("ambient(decontX) done=%s, cells=%d", ok, ncol(sce)))

File: workflow/scripts/py/qc_ambient.py
"""Stage 02 — ambient RNA (Python path: `none` passthrough or `cellbender`).

decontX / SoupX are handled by the R sibling (qc_ambient.R); they work on the FILTERED
matrices GEO ships. CellBender needs raw (unfiltered) droplets and a GPU to be practical;
when those aren't available it logs and passes through so the pipeline still completes.
"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from _utils import get_logger, read_h5ad, write_h5ad  # noqa: E402

log = get_logger(snakemake)  # noqa: F821
method = snakemake.params.method  # noqa: F821
adata = read_h5ad(snakemake.input.h5ad)  # noqa: F821

if method == "cellbender":
    log.warning(
        "cellbender requested via the Python path. CellBender needs the raw "
        "(unfiltered) droplet matrix and a GPU; for the GEO filtered matrices used here "
        "decontX (R path) is recommended. Passing counts through unchanged."
    )
adata.obs["ambient_fraction"] = 0.0
adata.uns["ambient_method"] = method
write_h5ad(adata, snakemake.output.h5ad)  # noqa: F821
log.info(f"ambient method={method}: wrote {adata.n_obs} cells")
